### PR TITLE
Add a full SFT GPU demo for Llama 3_1

### PR DESCRIPTION
**Pull request by katjasrz**
# Description

* Adds `src/MaxText/examples/sft_llama3_gpu.ipynb`: a GPU-focused, end-to-end notebook for SFT of Llama 3.1-8B on NVIDIA GPUs (HF auth → gated access note → HF → MaxText checkpoint conversion (CPU) → SFT run → TensorBoard → inference sanity check).
* Motivation: complements existing notebook/docs that emphasize TPU SFT flows (e.g. `sft_llama3_demo.ipynb`) with a clear NVIDIA GPU path.

# Tests

Executed the notebook in the NGC container `nvcr.io/nvidia/jax:26.01-maxtext-py3` on a cluster node with 8 H100 NVIDIA GPUs, CUDA 13.1, driver 580.105.08, JAX 0.8.1.dev20260217.

Verified:

* Checkpoint conversion completes.
* SFT runs for 100 steps and writes checkpoints/logs.
* TensorBoard logs are created.
* Inference sanity check runs and produces output.

To reproduce, execute as follows:

```bash
cd <YOUR_ROOT_DIRECTORY>

# clone this repo
git clone https://github.com/katjasrz/maxtext.git

export ROOT_DIR=$(pwd)
export PROJECT_DIR=$ROOT_DIR/maxtext/src/maxtext/examples
export HF_CACHE_DIR=$ROOT_DIR/huggingface

docker run -it --rm --ipc=host \
  --gpus=all \
  -p 8889:8889 \
  -p 6006:6006 \
  --shm-size=16g \
  --ulimit memlock=-1 \
  -v "$PROJECT_DIR":/workspace \
  -v "$HF_CACHE_DIR":/hf_cache \
  -e HF_HOME=/hf_cache \
  -e LOCAL_UID=$(id -u) \
  -e LOCAL_GID=$(id -g) \
  nvcr.io/nvidia/jax:26.01-maxtext-py3 \
  bash -lc 'set -e
    groupadd -g $LOCAL_GID hostgrp 2>/dev/null || true
    useradd -u $LOCAL_UID -g $LOCAL_GID -M -d /workspace hostusr 2>/dev/null || true
    
    python3 -m pip install --upgrade pip
    pip install jupyterlab ipywidgets
    pip install -U git+https://github.com/google/tunix
		pip install torch --index-url https://download.pytorch.org/whl/cpu
    
    su hostusr -c "cd /workspace && HOME=/workspace HF_HOME=/hf_cache \
      jupyter lab --ip=0.0.0.0 --port=8889 --no-browser"'
```

Then follow the instructions in the jupyter notebook src/maxtext/examples/sft_llama3_gpu.ipynb

Two container-specific fixes corresponding to an older maxtext version excluded from the notebook:

Fix 1. The code below includes a workaround for a known container issue where create_nnx_model defaults model_mode to None instead of "train". This is patched at runtime.

```bash
# Fix for container bug: model_creation_utils.create_nnx_model defaults model_mode=None
# but it should default to "train". Set the correct default.
from MaxText import model_creation_utils
model_creation_utils.create_nnx_model.__defaults__ = (None, None, "train", None)
```

Fix 2. The code below is a workaround for a known container issue where empty-string defaults for hf_train_files/hf_eval_files/hf_data_dir cause datasets.load_dataset to fail. These are patched to None at runtime.

```bash
# Fix for container bug: empty string defaults for hf_train_files/hf_eval_files/hf_data_dir
# cause datasets.load_dataset to fail. Monkey-patch to convert empty strings to None.
# Guard against multiple applications to avoid recursion.
import datasets
if not hasattr(datasets, '_original_load_dataset'):
    datasets._original_load_dataset = datasets.load_dataset

    def _patched_load_dataset(*args, **kwargs):
        for key in ['data_files', 'data_dir']:
            if key in kwargs and kwargs[key] == '':
                kwargs[key] = None
        return datasets._original_load_dataset(*args, **kwargs)

    datasets.load_dataset = _patched_load_dataset
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by katjasrz** [comment]
I've submitted the CLA.

File: src/maxtext/examples/sft_llama3_gpu.ipynb [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "93a0030b-ab25-4d76-a173-1a464c19087a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Prefer environment variable if already set\n",
+    "HF_TOKEN = os.environ.get(\"HF_TOKEN\")\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_gpu.ipynb]: Could you please add the details/formatting similar to as `Get Your Hugging Face Token` from `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb`? It is helpful for first time users and also for consistency

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    "print(f\"Number of available devices: {jax.local_device_count()}\")\n",
+    "print(f\"MaxText installation path: {MAXTEXT_REPO_ROOT}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1687aa03-1549-429a-8156-571c7493ca3d",
+   "metadata": {},
+   "source": [
+    "### Define model paths and run configuration\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_demo_gpu.ipynb:262]: just curious - do we need `jax.distributed.is_initialized()` like 
`https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb` here?

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+    "\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_demo_gpu.ipynb:263]: For consistency, how much of code cell [5] of `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb` do you think is applicable here?

File: src/maxtext/examples/sft_llama3_gpu.ipynb [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+    "\n",
+    "BASE_OUTPUT_DIRECTORY = Path(os.environ.get(\"BASE_OUTPUT_DIRECTORY\", str(WORKSPACE_DIR / \"sft_llama3_output\")))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6b762437-1edb-4123-8257-90cb98028e97",
+   "metadata": {},
+   "source": [
+    "### Download and convert the Llama 3.1 checkpoint\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_gpu.ipynb]: For consistency, could you please add your better logging/documentation to `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb`?

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+    "if not ckpt_dir.exists():\n",
+    "    raise RuntimeError(\"Model checkpoint conversion failed. See logs above.\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "265e9555-f026-4bbc-a6fb-7b0cac6bd9da",
+   "metadata": {},
+   "source": [
+    "## Provide the training configuration\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_demo_gpu.ipynb:386]: For consistency, could you please add your better logging/documentation to `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb`?

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+    "\n",
+    "This block builds the full MaxText SFT training configuration by loading the base `sft.yml` config and applying runtime overrides for the model, dataset, hyperparameters, and output paths. Each run is tagged with a timestamp-based name to keep outputs isolated across experiments. Key settings:\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_demo_gpu.ipynb:388]: For consistency, could you please add your better logging/documentation to `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb`?

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+    "print(f\"  Max sequence length: {config.max_target_length}\")\n",
+    "print(f\"  Output Directory: {config.base_output_directory}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "408c6100-20e9-4dcb-b9b6-42d68bb03ae7",
+   "metadata": {},
+   "source": [
+    "## Run the SFT training\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_demo_gpu.ipynb:452]: For consistency, could you please add your better logging/documentation to `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb`?

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+    "    print(f\"Error details: {e}\")\n",
+    "    raise"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "24e3a3e2-027a-4bb7-bb7f-163605010d03",
+   "metadata": {},
+   "source": [
+    "## Visualize training metrics with TensorBoard\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_demo_gpu.ipynb:495]: For consistency, could you please add this visualization snippet to `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb`?

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.12.3"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 5
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_demo_gpu.ipynb:602]: nit: please delete the last cell

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5acaf26b-72fe-404b-827a-297045547f5b",
+   "metadata": {},
+   "source": [
+    "## Test inference\n",
+    "\n",
+    "A quick sanity check to verify the fine-tuned model produces coherent output. The code below tokenizes a prompt using the Llama 3.1 chat template, then runs greedy autoregressive generation for up to 10 tokens, stopping early if the model produces an EOS token. This confirms the model loaded correctly and produces reasonable predictions.\n",
```
Comment (A9isha) [inline src/maxtext/examples/sft_llama3_demo_gpu.ipynb:514]: For consistency, could you please add this sampling snippet to `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb`?

**Review by A9isha — COMMENTED** [review]
This is awesome - thank you so much!


Also for consistency, could you please rename this to `maxtext/src/maxtext/examples
/sft_llama3_demo_gpu.ipynb`, and rename`https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo.ipynb` to `https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/examples/sft_llama3_demo_tpu.ipynb`?

And delete the `src/MaxText/examples/sft_llama3_gpu.ipynb` since `src/MaxText` is going to get deprecated in favor of `src/maxtext`

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb (hunk 2 above)
Comment (katjasrz) [reply src/maxtext/examples/sft_llama3_demo_gpu.ipynb:262]: I've added a section explaining how it works on different GPU setups

File: src/maxtext/examples/sft_llama3_demo_gpu.ipynb (hunk 3 above)
Comment (katjasrz) [reply src/maxtext/examples/sft_llama3_demo_gpu.ipynb:263]: This implementation is based on the original inline conversion snippet and preserves its core behavior, but improves structure and robustness by wrapping the logic in a reusable function, adding clearer error handling, controlled logging, and safer output directory handling. I’ve also moved the torch installation step into the setup description above, so users can install it once ahead of time and avoid triggering unnecessary reinstallation if it’s already present.

File: src/maxtext/examples/sft_llama3_gpu.ipynb (hunk 4 above)
Comment (katjasrz) [reply src/maxtext/examples/sft_llama3_gpu.ipynb]: Sure!
